File: Pacode/2026.2/turma_intermediaria/ListaIntermediaria_CoordinateCompression_SparseTable/G - Catapult that ball/G-catapult-that-ball.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n, m;
    while (cin >> n >> m) {
        vector<int> lg(n + 1);
        for (int i = 2; i <= n; i++) lg[i] = lg[i / 2] + 1;
        vector<vector<long long>> st(lg[n] + 1, vector<long long>(n));
        for (auto &x : st[0]) cin >> x;
        for (int k = 1; k <= lg[n]; k++) {
            int length = 1 << k;
            for (int i = 0; i + length <= n; i++) {
                st[k][i] = max(st[k - 1][i], st[k - 1][i + length / 2]);
            }
        }
        int answer = 0;
        while (m--) {
            int from, to;
            cin >> from >> to;
            int left = min(from, to), right = max(from, to);
            if (right - left <= 1) {
                answer++;
                continue;
            }
            int first = left, last = right - 2;
            int k = lg[last - first + 1], length = 1 << k;
            long long highest = max(st[k][first], st[k][last - length + 1]);
            if (highest <= st[0][from - 1]) answer++;
        }
        cout << answer << '\n';
    }
    return 0;
}

File: Pacode/2026.2/turma_intermediaria/ListaIntermediaria_CoordinateCompression_SparseTable/G - Catapult that ball/G-catapult-that-ball.py
import sys

data = iter(map(int, sys.stdin.buffer.read().split()))
answers = []

while True:
    try:
        n = next(data)
        m = next(data)
    except StopIteration:
        break

    heights = [next(data) for _ in range(n)]
    table = [heights]
    k = 1
    while 1 << k <= n:
        previous = table[-1]
        half = 1 << (k - 1)
        length = n - (1 << k) + 1
        table.append([max(previous[i], previous[i + half]) for i in range(length)])
        k += 1

    answer = 0
    for _ in range(m):
        start = next(data)
        end = next(data)
        left = min(start, end)
        right = max(start, end)
        if right - left <= 1:
            answer += 1
            continue

        first = left
        last = right - 2
        k = (last - first + 1).bit_length() - 1
        length = 1 << k
        highest = max(table[k][first], table[k][last - length + 1])
        if highest <= heights[start - 1]:
            answer += 1

    answers.append(str(answer))

sys.stdout.write('\n'.join(answers))
